### DISCUSSION TITLE
Handling error emitted from <ais-instantsearch> element

### DISCUSSION THREAD

**Discussion by ramikhafagi96** [algolia/instantsearch#5646] [Q&A] [answered]
We're using `<ais-instantsearch>` to fetch search hits from algolia search on our Angular application, for some connectivity issues on user devices we're getting this error on Sentry `Uncaught (in promise): RetryError: Unreachable hosts - your application id may be incorrect. If the error persists, contact support@algolia.com.`, this below is the stack trace and we need to know how to catch this error to avoid consuming our quota on Sentry, given that we're not calling the API using javascript code only by rendering the element in our template file.

```
./node_modules/instantsearch.js/es/lib/InstantSearch.js in call at line 408:14
./node_modules/@algolia/events/events.js in EventEmitter.prototype.emit at line 81:17
./node_modules/algoliasearch-helper/src/algoliasearch.helper.js in AlgoliaSearchHelper.prototype._dispatchAlgoliaError at line 1378:8
./node_modules/zone.js/dist/zone.js in _ZoneDelegate.prototype.invoke at line 412:30
./node_modules/@angular/core/fesm2020/core.mjs in zone._inner.onInvoke at line 24210:33
./node_modules/zone.js/dist/zone.js in _ZoneDelegate.prototype.invoke at line 411:56
./node_modules/zone.js/dist/zone.js in Zone.prototype.run at line 169:47
./node_modules/zone.js/dist/zone.js in apply at line 1329:38
./node_modules/zone.js/dist/zone.js in _ZoneDelegate.prototype.invokeTask at line 446:35
./node_modules/@angular/core/fesm2020/core.mjs in zone._inner.onInvokeTask at line 24197:33
./node_modules/zone.js/dist/zone.js in _ZoneDelegate.prototype.invokeTask at line 445:64
./node_modules/zone.js/dist/zone.js in Zone.prototype.runTask at line 214:51
./node_modules/zone.js/dist/zone.js in drainMicroTaskQueue at line 635:39
./node_modules/zone.js/dist/zone.js in y.invokeTask at line 532:25
./node_modules/zone.js/dist/zone.js in this.invoke at line 517:52
./node_modules/zone.js/dist/zone.js in S.args[0] at line 3138:36
./node_modules/zone.js/dist/zone.js in resolvePromise at line 1265:35
./node_modules/zone.js/dist/zone.js in apply at line 1336:21
./node_modules/zone.js/dist/zone.js in _ZoneDelegate.prototype.invokeTask at line 446:35
./node_modules/@angular/core/fesm2020/core.mjs in zone._inner.onInvokeTask at line 24197:33
./node_modules/zone.js/dist/zone.js in _ZoneDelegate.prototype.invokeTask at line 445:64
./node_modules/zone.js/dist/zone.js in Zone.prototype.runTask at line 214:51
./node_modules/zone.js/dist/zone.js in drainMicroTaskQueue at line 635:39
./node_modules/zone.js/dist/zone.js in y.invokeTask at line 532:25
./node_modules/zone.js/dist/zone.js in this.invoke at line 517:52
./node_modules/zone.js/dist/zone.js in S.args[0] at line 3138:36
```
``` javascript
// ./node_modules/instantsearch.js/es/lib/InstantSearch.js in call at line 408:14

// typescript lies here, error is in some cases { name: string, message: string }
          var err = error;
          error = Object.keys(err).reduce(function (acc, key) {

            acc[key] = err[key];
            return acc;
          }, new Error(err.message));
        }
        // If an error is emitted, it is re-thrown by events. In previous versions
        // we emitted {error}, which is thrown as:
        // "Uncaught, unspecified \"error\" event. ([object Object])"
        // To avoid breaking changes, we make the error available in both
```

```javascript
// ./node_modules/@algolia/events/events.js in EventEmitter.prototype.emit at line 81:17
// fast cases
      case 1:
        handler.call(this);
        break;
      case 2:
       -> handler.call(this, arguments[1]);
        break;
      case 3:
        handler.call(this, arguments[1], arguments[2]);
        break;
      // slower

```
```javascript
// ./node_modules/algoliasearch-helper/src/algoliasearch.helper.js in AlgoliaSearchHelper.prototype._dispatchAlgoliaError at line 1378:8
  }
  this._currentNbQueries -= queryId - this._lastQueryIdReceived;
  this._lastQueryIdReceived = queryId;
  this.emit('error', {
    error: error
  });
  if (this._currentNbQueries === 0) this.emit('searchQueueEmpty');
};
```

**Comment by Haroenv** [comment]
You can handle the errors yourself if you don't want them to be thrown. If you want to display the message that would be like this:

```vue
<ais-state-results catch-error>
  <template v-slot="{ status, error }">
    <p v-if="status === 'error'">{{ status }}: <q>{{ error.message }}</q></p>
    <span v-else />
  </template>
</ais-state-results>
```

Or if you only want to suppress the error without displaying the message:

```vue
<ais-state-results catch-error>
  <template>
    <span />
  </template>
</ais-state-results>
```

Here's an example: https://codesandbox.io/s/gallant-nobel-wbgwts?file=/src/App.vue:768-1286

Catch-error is documented here: https://www.algolia.com/doc/api-reference/widgets/state-results/vue/#widget-param-catch-error

**Reply by ramikhafagi96** [reply]
@Haroenv I looked up for this element in Angular documentation but it doesn't exist, is there another way in angular?

**Reply by Haroenv** [reply, marked as answer]
This repository is not for angular-instantsearch, however you can create a custom widget following the regular guide, and do `parentInstantSearch.on('error', () => {})`

**Reply by northern-mindset** [reply]
Why isn't this widget available in the InstantSearch.js flavour? Would be very useful. Just curious about the reasoning.

**Reply by Haroenv** [reply]
The main use case for this widget was seen as a way to conditionally render other widgets. As InstantSearch.js doesn't have truly a rendering system that involves other widgets, that wasn't seen as useful. As well as that, the DX of a custom widget wasn't seen as an improvement over an event handler. Hope that makes sense @anantakrishna. 

As you discovered in the other widget as well, the widget life cycle isn't the same as the event life cycle, and it would require us to make changes in how events and renders work to be able to render on error too.

**Reply by northern-mindset** [reply]
Thanks a lot for the explanation, makes sense.

> the DX of a custom widget wasn't seen as an improvement over an event handler

This is particularly interesting. I guess here you mean that instead of a custom widget one could just add some event listeners to the search instance and do all the stuff in handlers. However, what about re-usability and unification? The widget system adds a well-defined protocol of interacting between the DOM and the Algolia API. Wouldn't like to invent another protocol here…

**Reply by Haroenv** [reply]
That's a good point, overall I'd say that displaying errors isn't an area that has had significant investment, we might indeed come back on that solution later and decide that widgets _should_ render for errors explicitly, but that has significant other effects (need error handling in every widget, for example).

Hope that helps, and that you're happy with the design of your error handler now :)
